File: src/core/wasm/__tests__/CablePhysics.test.ts
import { describe, it } from "node:test";
import assert from "node:assert";
import { CablePhysics } from "../CablePhysics";
import { SpatialIndex } from "../SpatialIndex";
import { SplineSolver } from "../SplineSolver";

describe("Sub-Phase 4.2: Verlet Cable Physics & Spatial Index", () => {
  it("initializes cable particles and pins endpoints", () => {
    const cable = new CablePhysics("exec", 16);
    cable.initialize({ x: 50, y: 50 }, { x: 350, y: 50 });

    const particles = cable.getParticles();
    assert.strictEqual(particles.length, 16);
    assert.strictEqual(particles[0].isPinned, true);
    assert.strictEqual(particles[15].isPinned, true);
    assert.strictEqual(particles[0].position.x, 50);
    assert.strictEqual(particles[15].position.x, 350);
  });

  it("settles 200 simultaneous wires within 300ms of a node drag without visible jitter", () => {
    const WIRE_COUNT = 200;
    const wires: CablePhysics[] = [];

    for (let i = 0; i < WIRE_COUNT; i++) {
      const type = i % 2 === 0 ? "exec" : "number";
      const wire = new CablePhysics(type, 16);
      wire.initialize({ x: 0, y: i * 10 }, { x: 300, y: i * 10 + 40 });
      wires.push(wire);
    }

    // Simulate node drag moving endpoint
    for (const wire of wires) {
      wire.setEndpoints({ x: 0, y: 0 }, { x: 450, y: -30 });
    }

    // Step for 300ms at 60 FPS (18 frames)
    const dt = 1 / 60;
    const steps = 18;
    const t0 = Date.now();

    for (let step = 0; step < steps; step++) {
      for (const wire of wires) {
        wire.step(dt);
      }
    }

    const elapsedMs = Date.now() - t0;
    let settledCount = 0;
    for (const wire of wires) {
      if (wire.isSettled(1.5)) {
        settledCount++;
      }
    }

    // At least 95% of cables settle within 300ms
    assert.ok(settledCount >= 190, `Expected >= 190 settled wires, got ${settledCount}`);
    assert.ok(elapsedMs < 1000, `Simulation too slow: ${elapsedMs}ms`);
  });

  it("enforces tunable stiffness profiles (exec wires stiffer than data wires)", () => {
    const execWire = new CablePhysics("exec", 16);
    execWire.initialize({ x: 0, y: 100 }, { x: 400, y: 100 });

    const dataWire = new CablePhysics("string", 16);
    dataWire.initialize({ x: 0, y: 100 }, { x: 400, y: 100 });

    // Step 60 frames
    for (let i = 0; i < 60; i++) {
      execWire.step(1 / 60);
      dataWire.step(1 / 60);
    }

    const execMidY = execWire.getParticles()[8].position.y;
    const dataMidY = dataWire.getParticles()[8].position.y;

    // Exec mid-point should sag less (lower y value)
    assert.ok(
      execMidY < dataMidY,
      `Exec wire mid (${execMidY}) should have less sag than data wire mid (${dataMidY})`
    );
  });

  it("generates smooth SVG Catmull-Rom path from simulated particles", () => {
    const cable = new CablePhysics("default", 16);
    cable.initialize({ x: 100, y: 100 }, { x: 300, y: 200 });

    cable.step(1 / 60);
    cable.step(1 / 60);

    const svgPath = cable.generateSvgPath();
    assert.ok(svgPath.startsWith("M 100.0 100.0"));
    assert.ok(svgPath.includes(" C "));
  });

  it("snaps particles along Hermite spline on resetAlongCurve", () => {
    const cable = new CablePhysics("exec", 16);
    cable.initialize({ x: 0, y: 0 }, { x: 200, y: 100 });

    const spline = SplineSolver.calculateWireSpline({ x: 0, y: 0 }, { x: 200, y: 100 });
    cable.resetAlongCurve(spline);

    const particles = cable.getParticles();
    assert.strictEqual(particles[0].position.x, 0);
    assert.strictEqual(particles[0].position.y, 0);
    assert.strictEqual(particles[15].position.x, 200);
    assert.strictEqual(particles[15].position.y, 100);
  });

  it("SpatialIndex: Quadtree indexes nodes with O(log N) point and range queries", () => {
    const index = new SpatialIndex({ minX: 0, minY: 0, maxX: 4000, maxY: 4000 });

    // Insert 100 nodes
    for (let i = 0; i < 100; i++) {
      const x = (i % 10) * 350 + 50;
      const y = Math.floor(i / 10) * 350 + 50;
      index.insert(`node_${i}`, "node", { minX: x, minY: y, maxX: x + 240, maxY: y + 160 }, i);
    }

    assert.strictEqual(index.size(), 100);

    // Hit-test at (60, 60)
    const hit = index.hitTestTopmost(60, 60);
    assert.ok(hit !== null);
    assert.strictEqual(hit.id, "node_0");

    // Marquee range query
    const inRange = index.queryRange({ minX: 0, minY: 0, maxX: 800, maxY: 800 });
    assert.ok(inRange.length >= 4);

    // Incremental drag update
    index.update("node_0", { minX: 2000, minY: 2000, maxX: 2240, maxY: 2160 });
    assert.strictEqual(index.hitTestTopmost(60, 60), null);
    const movedHit = index.hitTestTopmost(2010, 2010);
    assert.ok(movedHit !== null);
    assert.strictEqual(movedHit.id, "node_0");

    // Remove
    const removed = index.remove("node_0");
    assert.strictEqual(removed, true);
    assert.strictEqual(index.size(), 99);
    assert.strictEqual(index.hitTestTopmost(2010, 2010), null);
  });
});

File: wasm/tests/CablePhysics.test.cpp
#include "../include/CablePhysics.hpp"
#include "../include/SpatialIndex.hpp"
#include <iostream>
#include <cassert>
#include <chrono>

using namespace WebAppEngine;

void testCablePhysicsInitialization() {
    CablePhysics cable(WireType::Exec);
    cable.initialize(Point2D(100.0f, 100.0f), Point2D(400.0f, 100.0f), CablePhysics::getConfigForWireType(WireType::Exec));

    const auto& particles = cable.getParticles();
    assert(particles.size() == 16);
    assert(particles.front().isPinned);
    assert(particles.back().isPinned);
    assert(particles.front().position.x == 100.0f);
    assert(particles.back().position.x == 400.0f);

    std::cout << "✔ CablePhysics: Particle initialization and pinning verified\n";
}

void testSettlingTimeAndOvershoot() {
    // 200 simultaneous wires settle within 300ms of a node drag without visible jitter
    const int WIRE_COUNT = 200;
    std::vector<CablePhysics> wires;
    wires.reserve(WIRE_COUNT);

    for (int i = 0; i < WIRE_COUNT; ++i) {
        WireType type = (i % 2 == 0) ? WireType::Exec : WireType::DataNumber;
        CablePhysics wire(type);
        wire.initialize(Point2D(0.0f, static_cast<float>(i * 10)),
                        Point2D(300.0f, static_cast<float>(i * 10 + 50)),
                        CablePhysics::getConfigForWireType(type));
        wires.push_back(wire);
    }

    // Simulate node drag event by shifting endpoint by (150px, -80px)
    for (auto& wire : wires) {
        wire.setEndpoints(Point2D(0.0f, 0.0f), Point2D(450.0f, -30.0f));
    }

    // Step simulation at 60 FPS (dt = 0.0166s = 16.6ms) for 300ms (18 steps)
    const float dt = 1.0f / 60.0f;
    const int stepsFor300ms = 18; // ~300ms
    auto startTime = std::chrono::high_resolution_clock::now();

    for (int step = 0; step < stepsFor300ms; ++step) {
        for (auto& wire : wires) {
            wire.step(dt);
        }
    }

    auto endTime = std::chrono::high_resolution_clock::now();
    auto elapsedUs = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime).count();

    int settledCount = 0;
    for (const auto& wire : wires) {
        if (wire.isSettled(1.5f)) {
            settledCount++;
        }
    }

    assert(settledCount >= 195);
    std::cout << "✔ CablePhysics: 200 simultaneous wires settled within 300ms ("
              << settledCount << "/200 settled, simulation computed in " << elapsedUs / 1000.0f << " ms)\n";
}

void testWireTypeDifferences() {
    CablePhysics execWire(WireType::Exec);
    execWire.initialize(Point2D(0.0f, 100.0f), Point2D(400.0f, 100.0f), CablePhysics::getConfigForWireType(WireType::Exec));

    CablePhysics dataWire(WireType::DataString);
    dataWire.initialize(Point2D(0.0f, 100.0f), Point2D(400.0f, 100.0f), CablePhysics::getConfigForWireType(WireType::DataString));

    // Step for 1 second of gravity settling
    for (int i = 0; i < 60; ++i) {
        execWire.step(1.0f / 60.0f);
        dataWire.step(1.0f / 60.0f);
    }

    // Exec wire should have significantly less sag than data wire
    float execMidY = execWire.getParticles()[8].position.y;
    float dataMidY = dataWire.getParticles()[8].position.y;

    assert(execMidY < dataMidY); // Exec is higher (less saggy)
    std::cout << "✔ CablePhysics: Exec wires stiffer with less sag than data wires (Exec: "
              << execMidY << "px vs Data: " << dataMidY << "px)\n";
}

void testSpatialIndexQuadtree() {
    SpatialIndex index(AABB(0.0f, 0.0f, 4000.0f, 4000.0f));

    // Insert 100 nodes across canvas
    for (int i = 0; i < 100; ++i) {
        float x = static_cast<float>((i % 10) * 350 + 50);
        float y = static_cast<float>((i / 10) * 350 + 50);
        index.insert("node_" + std::to_string(i), SpatialItemType::NodeCard, AABB(x, y, x + 240.0f, y + 160.0f), i);
    }

    assert(index.size() == 100);

    // Test Point Query (Hit-test)
    SpatialItem hitItem;
    bool hit = index.hitTestTopmost(60.0f, 60.0f, hitItem);
    assert(hit);
    assert(hitItem.id == "node_0");

    // Test Range Query (Marquee selection in [0, 0] to [800, 800])
    auto selected = index.queryRange(AABB(0.0f, 0.0f, 800.0f, 800.0f));
    assert(selected.size() >= 4);

    // Test Incremental Drag Update
    index.update("node_0", AABB(2000.0f, 2000.0f, 2240.0f, 2160.0f));
    assert(!index.hitTestTopmost(60.0f, 60.0f, hitItem)); // No longer at old spot
    assert(index.hitTestTopmost(2010.0f, 2010.0f, hitItem)); // Found at new spot
    assert(hitItem.id == "node_0");

    // Test Removal
    bool removed = index.remove("node_0");
    assert(removed);
    assert(index.size() == 99);
    assert(!index.hitTestTopmost(2010.0f, 2010.0f, hitItem));

    std::cout << "✔ SpatialIndex: Quadtree hit-test, incremental update, and range query verified\n";
}

int main() {
    std::cout << "========================================================\n";
    std::cout << "RUNNING CABLE PHYSICS & SPATIAL INDEX VERIFICATION TESTS\n";
    std::cout << "========================================================\n";

    testCablePhysicsInitialization();
    testSettlingTimeAndOvershoot();
    testWireTypeDifferences();
    testSpatialIndexQuadtree();

    std::cout << "All Sub-Phase 4.2 C++ unit tests passed successfully!\n";
    return 0;
}
